src/Processors/Formats/Impl/ParquetBloomFilterCondition.cpp

```diff
@@ -0,0 +1,433 @@
+#include <DataTypes/DataTypeLowCardinality.h>
+#include <Interpreters/misc.h>
+#include <Interpreters/convertFieldToType.h>
+#include <Columns/ColumnConst.h>
+
+#include <parquet/bloom_filter.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int LOGICAL_ERROR;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+namespace
+{
+
+parquet::ByteArray createByteArray(std::string_view view, TypeIndex type, uint8_t * buffer, uint32_t buffer_size)
+{
+    if (isStringOrFixedString(type))
+    {
+        return view;
+    }
+    else
+    {
+        auto size = static_cast<uint32_t>(std::max(view.size(), sizeof(uint32_t)));
+        chassert(size <= buffer_size);
+        std::copy(view.begin(), view.end(), buffer);
+        return parquet::ByteArray(size, buffer);
+    }
+}
+
+bool maybeTrueOnBloomFilter(const IColumn * data_column, const std::unique_ptr<parquet::BloomFilter> & bloom_filter, bool match_all)
+{
+    static constexpr uint32_t buffer_size = 32;
+    uint8_t buffer[buffer_size] = {0};
+
+    auto column_size = data_column->size();
+
+    for (size_t i = 0; i < column_size; ++i)
+    {
+        const auto data_view = data_column->getDataAt(i).toView();
+
+        const auto ba = createByteArray(data_view, data_column->getDataType(), buffer, buffer_size);
+        bool found = bloom_filter->FindHash(bloom_filter->Hash(&ba));
+
+        if (match_all && !found)
+            return false;
+        if (!match_all && found)
+            return true;
+    }
+
+    return match_all;
+}
+
+DataTypePtr getPrimitiveType(const DataTypePtr & data_type)
+{
+    if (const auto * array_type = typeid_cast<const DataTypeArray *>(data_type.get()))
+    {
+        if (!typeid_cast<const DataTypeArray *>(array_type->getNestedType().get()))
+            return getPrimitiveType(array_type->getNestedType());
+        else
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unexpected type {} of bloom filter index.", data_type->getName());
+    }
+
+    if (const auto * nullable_type = typeid_cast<const DataTypeNullable *>(data_type.get()))
+        return getPrimitiveType(nullable_type->getNestedType());
+
+    if (const auto * low_cardinality_type = typeid_cast<const DataTypeLowCardinality *>(data_type.get()))
+        return getPrimitiveType(low_cardinality_type->getDictionaryType());
+
+    return data_type;
+}
+
+ColumnWithTypeAndName getPreparedSetInfo(const ConstSetPtr & prepared_set)
+{
+    if (prepared_set->getDataTypes().size() == 1)
+        return {prepared_set->getSetElements()[0], prepared_set->getElementsTypes()[0], "dummy"};
+
+    Columns set_elements;
+    for (auto & set_element : prepared_set->getSetElements())
+
+        set_elements.emplace_back(set_element->convertToFullColumnIfConst());
+
+    return {ColumnTuple::create(set_elements), std::make_shared<DataTypeTuple>(prepared_set->getElementsTypes()), "dummy"};
+}
+
+}
+
+ParquetBloomFilterCondition::ParquetBloomFilterCondition(
```

Comment:
This class is heavily inspired on https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeIndexBloomFilter.cpp (formerly MergeTreeIndexConditionBloomFilter). Lots of code duplication.

Reply:
I am open for a refactoring, but not sure if this is a good idea and what's the best way to do it

Reply:
I would try to extend KeyCondition instead, so that it works for both min/max filters and bloom filters. Ideally refactor MergeTreeConditionBloomFilterText and MergeTreeIndexConditionBloomFilter to use it too, if you want.

So, add things like FUNCTION_HAS to KeyCondition. Instead of FUNCTION_EQUALS use FUNCTION_IN_RANGE with single-element range. Ignore irrelevant conditions like FUNCTION_IN_RANGE with non-signular range (as if they were FUNCTION_UNKNOWN).

(If there are lots of irrelevant conditions to ignore, and that slows down the query, the KeyCondition can have a method to specialize it for bloom filtering by collapsing irrelevant conditions into FUNCTION_UNKNOWN. So, setKeyConditionImpl() would build a KeyCondition with enough information for both types of filtering, then input format would optionally trim it down for specific tasks. But probably it'll be fast enough without that.)

I didn't look at this code in detail, maybe there's some reason this wouldn't work. If you try it and it doesn't work, please let me know :)